### PR TITLE
🔴 🚨  Resizing tokens embeddings: initialize from old embeddings' normal distribution.

### LINKED ISSUE #32948: `resize_token_embeddings` in NLLB leading to empty outputs
### System Info

- `transformers` version: 4.42.3
- Platform: Linux-4.18.0-513.11.1.el8_9.x86_64-x86_64-with-glibc2.28
- Python version: 3.10.14
- Huggingface_hub version: 0.23.4
- Safetensors version: 0.4.3
- Accelerate version: 0.33.0
- Accelerate config:    not found
- PyTorch version (GPU?): 2.3.1+cu121 (False)
- Tensorflow version (GPU?): not installed (NA)
- Flax version (CPU?/GPU?/TPU?): not installed (NA)
- Jax version: not installed
- JaxLib version: not installed
- Using distributed or parallel set-up in script?: yes

### Who can help?

@ArthurZucker

### Information

- [X] The official example scripts
- [ ] My own modified scripts

### Tasks

- [X] An officially supported task in the `examples` folder (such as GLUE/SQuAD, ...)
- [ ] My own task or dataset (give details below)

### Reproduction

```
from transformers import AutoModelForSeq2SeqLM, AutoTokenizer

model = AutoModelForSeq2SeqLM.from_pretrained("facebook/nllb-200-distilled-600M")
tokenizer = AutoTokenizer.from_pretrained("facebook/nllb-200-distilled-600M", additional_special_tokens=[f"code_{i}" for i in range(18)], use_fast=True)
model.resize_token_embeddings(len(tokenizer))
```

After resizing, generation using an official example:
```
article = "Şeful ONU spune că nu există o soluţie militară în Siria"
inputs = tokenizer(article, return_tensors="pt")

translated_tokens = model.generate(
    **inputs, forced_bos_token_id=tokenizer.convert_tokens_to_ids("deu_Latn"), max_length=30
)
tokenizer.batch_decode(translated_tokens, skip_special_tokens=True)[0]
```
Output is: `'t t t t t t t t t t'`

### Expected behavior

Generation should work without any errors. One interesting thing to note here is if I add just 2 new tokens, it works fine.

### PR DESCRIPTION
**Pull request by abuelnasr0**
# What does this PR do?

This PR initializes new tokens embeddings from a normal distribution with the old embeddings' mean and covariance. as described in this article https://nlp.stanford.edu/~johnhew/vocab-expansion.html. Thanks to this article, you can now add new tokens to your model without affecting its generation accuracy. 

Fixes #32948 

## Before submitting
- [ ] This PR fixes a typo or improves the docs (you can dismiss the other checks if that's the case).
- [x] Did you read the [contributor guideline](https://github.com/huggingface/transformers/blob/main/CONTRIBUTING.md#create-a-pull-request),
      Pull Request section?
- [ ] Was this discussed/approved via a Github issue or the [forum](https://discuss.huggingface.co/)? Please add a link
      to it if that's the case.
- [ ] Did you make sure to update the documentation with your changes? Here are the
      [documentation guidelines](https://github.com/huggingface/transformers/tree/main/docs), and
      [here are tips on formatting docstrings](https://github.com/huggingface/transformers/tree/main/docs#writing-source-documentation).
- [ ] Did you write any new necessary tests?

## Who can review?

cc: @ArthurZucker @LysandreJik

### CODE REVIEW

**Comment by abuelnasr0** [comment]
This gist shows the before and after GPT2 model generation after adding new tokens: https://colab.research.google.com/gist/abuelnasr0/7cc8decff72ccdbcf5073ad8259ee360/embedding_resize.ipynb

**Comment by Rocketknight1** [comment]
Yes, I'm happy to take it, it seems like a great PR!

**Review by Rocketknight1 — APPROVED** [review]
My overall impression is that this is a great addition that we should definitely merge, because the existing behaviour is highly undesirable, as mentioned in the [article](https://nlp.stanford.edu/~johnhew/vocab-expansion.html). My issues are basically nits:

- We have a mild preference for avoiding "math variable" naming, though I don't want to bloat the code either. Maybe replace `mu` and `cov` with `mean_embedding` and `covariance` or `covariance_matrix`?
- It'd be great if we could add a small test for this. There is a `test_resize_tokens_embedding` in `tests/test_modeling_common.py`, and the test could either be appended to that, or as a new test just below it. The test could check that new tokens are relatively close to the mean of the old tokens, though please set the tolerances quite loose - it'll be a real pain if the test becomes flaky because an outlier value is sampled!

**Comment by abuelnasr0** [comment]
Thanks @LysandreJik 
And thanks @Rocketknight1 for the review and the comments. I have added a test case to check if the new embeddings' mean is close to the original embeddings' mean, I think that will reduce the effect of any outlier values.

File: tests/test_modeling_common.py [hunk 1]
```diff
@@ -1782,6 +1782,10 @@ def test_resize_tokens_embeddings(self):
             self.assertEqual(type_model_embed_pre_resize, type_model_embed_post_resize)
             # Check that the model can still do a forward pass successfully (every parameter should be resized)
             model(**self._prepare_for_class(inputs_dict, model_class))
+            # Check that added embeddings mean is close to the old embeddings mean
+            old_embeddings_mean = torch.mean(model_embed.weight.data[:model_vocab_size, :], axis=0)
+            new_embeddings_mean = torch.mean(model_embed.weight.data[-10:, :], axis=0)
```
Comment (Rocketknight1) [inline tests/test_modeling_common.py:1787]:
```suggestion
            old_embeddings_mean = torch.mean(model_embed.weight.data[:-10, :], axis=0)
            new_embeddings_mean = torch.mean(model_embed.weight.data[-10:, :], axis=0)
```
For consistency, let's use -10 in both places!

**Review by Rocketknight1 — APPROVED** [review]
LGTM! One final extra-minor nit, but I'm happy with it now. cc @lysandrejik for final review

File: src/transformers/modeling_utils.py [hunk 2]
```diff
@@ -2162,8 +2162,24 @@ def _get_resized_embeddings(
             dtype=old_embeddings.weight.dtype,
         )
 
-        # initialize all new embeddings (in particular added tokens)
-        self._init_weights(new_embeddings)
+        # initialize new embeddings (in particular added tokens) if `new_num_tokens` is larger
```
Comment (ArthurZucker) [inline src/transformers/modeling_utils.py:2165]: one thing we need to make sure of is that this does not have issue with deepspeed and multinodes! (as for example the mean would require an all gather I think. This computation has to be done on gpu 0 is what my intuition tells me)

**Review by ArthurZucker — COMMENTED** [review]
Thanks! I think we need a test for `test_resize_tokens_embeddings` with deepspeed and multi GPU as we had a lot of breaking changes in the past year around it! #26102 #32192 #21065 etc etc

File: src/transformers/modeling_utils.py (hunk 2 above)
Comment (abuelnasr0) [reply src/transformers/modeling_utils.py:2165]: That is absolutely right, thanks for mentioning that. I have added support for Deepspeed and tested it manually. I will add test cases for it tomorrow!

Comment (abuelnasr0) [reply src/transformers/modeling_utils.py:2165]: @ArthurZucker I have added tests for deepspeed but they fail because the test env doesn't have deepspeed installed. what could be a solution for this? should I add the tests to `transformers/tests/deepspeed/test_deepspeed.py` for example?

Comment (Rocketknight1) [reply src/transformers/modeling_utils.py:2165]: @abuelnasr0 when tests depend on a library like deepspeed, tag them with the `@require_deepspeed` decorator so the test runner knows how to handle them! You can search the codebase for other examples where it's used and just copy the imports/decorator from there.

Comment (abuelnasr0) [reply src/transformers/modeling_utils.py:2165]: @Rocketknight1 Thank you for your response. It turns out that I was importing Deepspeed without checking if it was available. I fixed it.
@Rocketknight1 @ArthurZucker The code is ready for review now. Could you check it and see if the tests are all right?

Comment (ArthurZucker) [reply src/transformers/modeling_utils.py:2165]: Sorry for the delay!

**Review by ArthurZucker — APPROVED** [review]
Looks great to me, it's abreaking change so let's make sure we warn users about it!

File: src/transformers/modeling_utils.py (hunk 2 above)
Comment (abuelnasr0) [reply src/transformers/modeling_utils.py:2165]: @ArthurZucker  I have added 🔴 to the PR title. Is that what you meant? or should I add a `logger.warning()` to the code describing the new change in initializing embedding weights?

Comment (abuelnasr0) [reply src/transformers/modeling_utils.py:2165]: >let's make sure we warn users about it!

I didn't see this comment the first time I opened the PR! I will add a warning to the user describing new changes.

**Review by ArthurZucker — COMMENTED** [review]
Okay looks great! Given how big of a change this is, let's add a flag to `_get_resized_embeddings`, something like `multivariate_resizing` WDYT?

**Comment by abuelnasr0** [comment]
@ArthurZucker Sorry for the delay. last week was very busy for me.
I have added the flag to `resize_token_embeddings`,  `_get_resized_embeddings`, and `_get_resized_lm_head`. 
`_get_resized_lm_head` is very important for models that have untied weights. This should have been added from the beginning but I don't know how I didn't notice that.

**Comment by abuelnasr0** [comment]
Bias for linear lm_heads is not concluded in the article. but I have came to the conclusion that initializing with zero is the best solution.
here is more information given the conclusion in the article:

$p_{\theta'}(w_i \mid w_{1:i-1}) = \frac{\exp(h_{i-1}^{\top} e_{w_i} + b_{w_i})}{Z + \exp(h_{i-1}^\top e_{n+1} +  b_{n+1} ) }$
$p_{\theta'}(w_i \mid w_{1:i-1}) = \frac{\exp(h_{i-1}^{\top} e_{w_i} + b_{w_i})}{Z + \exp(h_{i-1}^\top e_{n+1} ) exp( b_{n+1})}$

If the new bias equals zero, then exp(0) = 1, and then the proof in the article will remain the same.

**Comment by Rocketknight1** [comment]
@abuelnasr0 I haven't dived into the math, but that seems counter-intuitive to me. Most models do not have a bias on the output head, but if they do, I would guess that the bias is usually large and negative for most logits (I can't test this, though, because I can't find any examples of models with bias on the output head!)

If the average bias is large and negative, then initializing new tokens with the mean embedding but a zero bias will mean that logits for the new token will be very large, right?

**Comment by abuelnasr0** [comment]
@Rocketknight1 That's right. 
I didn't consider the new value of `Z` Which has also bias in it.

After considering the new value of `Z`, the new bias should be initialized with the old bias mean. I have tried to be more focused this time. I hope I am not wrong again :sweat_smile:

this link contains a proof that should be replaced with "Averaging bounds the KL-divergence" part in the article: https://imgur.com/a/ZZQ3Mwk

**Comment by abuelnasr0** [comment]
I have initialized the new bias like this:
```python
new_lm_head.bias.data.normal_(mean=bias_mean, std=bias_std * 1e-5)
```
I have multiplied the std by 1e-5 just because the article initializes the embeddings by multiplying the covariance by 1e-5. I don't really know why, maybe to make the new embeddings alot closer to the mean.

**Review by ArthurZucker — COMMENTED** [review]
Thanks, getting in a good shape! Let's help our users a little bit more, and make the code more readable !

File: src/transformers/modeling_utils.py [hunk 3]
```diff
@@ -2162,8 +2188,49 @@ def _get_resized_embeddings(
...
+            # initialize new embeddings (in particular added tokens) if `new_num_tokens` is larger
+            # than `old_num_tokens`. The new embeddings will be sampled from a multivariate normal
+            # distribution that has old embeddings' mean and covariance. as described in this article:
+            # https://nlp.stanford.edu/~johnhew/vocab-expansion.html
+            logger.warning(
+                "The new embeddings will be sampled from a multivariate normal distribution that has old embeddings' mean and covariance. "
+                "As described in this article: https://nlp.stanford.edu/~johnhew/vocab-expansion.html."
+            )
+            added_num_tokens = new_num_tokens - old_num_tokens
+            if is_deepspeed_zero3_enabled() and not is_quantized:
```
[the remark this answers is not in the document]
Comment (abuelnasr0) [reply src/transformers/modeling_utils.py:2202]: I have introduced three functions:
`self._init_added_embeddings_weights_with_mean()`
`self._init_added_lm_head_weights_with_mean()` and it uses `self._init_added_embeddings_weights_with_mean()`
`self._init_added_lm_head_bias_with_mean()`
This will improve code usability for our case. what do you think? I am open to any other change.

Also, I think that `mean_resizing` is more user-friendly and explains the whole point of the new resizing technique.

**Review by ArthurZucker — APPROVED** [review]
Cleaner! Thanks a lot for this update

**Comment by ArthurZucker** [comment]
The only thing left is to fix the failing test which are related (example torch for example) 🤗

**Comment by abuelnasr0** [comment]
@ArthurZucker Thanks for your review.
Regarding tests, all tests have passed after rebasing.

**Comment by ArthurZucker** [comment]
All good for me! 🤗

**Comment by Rocketknight1** [comment]
Congrats on the PR @abuelnasr0, and thank you!

**Comment by abuelnasr0** [comment]
@Rocketknight1 Thank you for the reviews and the help.
